[7.0.x] intermediate upgrades

a-palchikov opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
This PR ports support for intermediate upgrade paths from 5.5.x branch. I've removed the ability to generate the planet/teleport configuration packages as well as planet secrets using CLI interface of the intermediate gravity binary as it appears to be unnecessary for 7.0.x (when going from 5.5.x via 6.1.x). See a comment below for additional thoughts/concerns.

## Type of change
<!--Required. Keep only those that apply.-->

* New feature (non-breaking change which adds functionality)
* This change has a user-facing impact

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Refs https://github.com/gravitational/gravity/issues/1588
<!--This PR depends on the following PRs (e.g. planet, satellite, etc.).-->
* Requires https://github.com/gravitational/gravity.e/pull/4320, https://github.com/gravitational/planet/pull/714
<!--This PR is a back-/forward-port of the following PR.-->
* Ports https://github.com/gravitational/gravity/pull/637

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Write tests
- [x] Perform manual testing
- [ ] Write documentation
- [ ] Address review feedback

## Implementation
<!--Optional. Add any relevant implementation details that might help the reviewers.-->

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->

### 1. 3-node cluster (1 controller, 2 workers)
    1. Install `5.5.49` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation includes an intermediate hop via `6.1.31`
    1. Upgrade to this branch

### 2. 3-node cluster (3 controllers)
    1. Install `5.5.49` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation includes an intermediate hop via `6.1.31`
    1. Upgrade to this branch

### 3. 3-node cluster (1 controller, 2 workers)
    1. Install `6.1.30` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation does not include any intermediate hops and goes directly to this branch

### 3-node cluster (3 controllers)
    1. Install `6.1.30` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation does not include any intermediate hops and goes directly to this branch

## Additional information
<!--Optional. Anything else that may be relevant.-->
While working on the port, it occurred to me that in order to be truly generic w.r.t intermediate hops we need to be able to generate sub-plans for each intermediate hops and have the binary of any given version execute it both ways during the plan execution. I did not explore the idea too far but I feel this would be the next logical iteration that will enable co-operation of arbitrary versions without the fear that something might break when only the newest version is doing all the steps (as the case is now).

a-palchikov committed:
Port intermediate upgrade changes from 5.5.x

a-palchikov committed:
Complete port of changes to be able to compile

a-palchikov committed:
Remove fixmes and commented out code

a-palchikov committed:
Revert support for extensions.v1beta1.PodSecurityPolicy to support
intermediate upgrades from 5.5.x.
Temporarily disable use of offline generation for planet/teleport
configuration and secrets generation.

a-palchikov committed:
Remove FIXMEs. Wire up openEBS and SELinux steps

a-palchikov committed:
Fix cluster upgrade tests w.r.t intermediate upgrades

a-palchikov committed:
Rid of the intermediate internal package as 7.0.x branch does not require external tools to generate package configurations

a-palchikov committed:
* Fix an issue with rpc/server agent not handling error exits properly.
 * Describe kubernetes Job values correctly

a-palchikov committed:
Revert " * Fix an issue with rpc/server agent not handling error exits properly."

This reverts commit c0a3ba825c8a614e52d0010193601512e4306d7a.

a-palchikov committed:
Fix rebase

a-palchikov committed:
Update jenkinsfile to build with intermediate step

a-palchikov committed:
Revisit the way gravity cluster upgrade trigger looks for the new
application version and add version lookup tests.
Use the new lib/update/cluster/versions package to encapsulate the
functionality of verifying the upgrade path.

a-palchikov committed:
Remove bogus validatePreconditions in favor of the extension interface

a-palchikov commented on lib/app/sync.go:
```diff
@@ -0,0 +1,178 @@
+	// ImageService specifies the docker registry service
+	ImageService docker.ImageService
+	// Progress specifies the optional progress message printer
+	Progress utils.Printer
```
I removed the `ScanConfig` attribute since it was not being used anywhere.

a-palchikov commented on lib/fsm/kubernetes.go:
```diff
@@ -104,6 +105,21 @@ func GetUpsertBootstrapResourceFunc(client *kubernetes.Clientset) resources.Reso
 				return trace.Wrap(rigging.ConvertError(err))
 			}
 			log.Debugf("Updated PodSecurityPolicy %q.", resource.Name)
+		case *v1beta1.PodSecurityPolicy:
```
This has been reverted to support RBAC configurations from older versions when using intermediate hops.

a-palchikov commented on lib/loc/loc.go:
```diff
@@ -238,6 +234,24 @@ func Deduplicate(ls []Locator) (result []Locator) {
 
+// GetUpdatedDependencies compares installedDeps against the updateDeps
+// and returns only locators from updateDeps that are updates of those given with installedDeps
+func GetUpdatedDependencies(installedDeps, updateDeps []Locator) ([]Locator, error) {
```
Moved from lib/app from what used to be [GetUpdatedDependencies](https://github.com/gravitational/gravity/blob/1d5bd035e9cddefafaf5d9f1471f68e9589a3d45/lib/app/apps.go#L141)

a-palchikov commented on lib/ops/constants.go:
```diff
@@ -241,4 +243,11 @@ var (
+	//
+	// Github issue: https://github.com/gravitational/gravity/issues/1445.
+	// KB: https://community.gravitational.com/t/recover-teleport-nodes-failing-to-join-due-to-bad-token/649.
+	TeleportBrokenJoinTokenVersion = semver.New("5.5.40")
```
Moved from [here](https://github.com/gravitational/gravity/blob/1d5bd035e9cddefafaf5d9f1471f68e9589a3d45/lib/ops/opsservice/versions.go#L264) which now has been moved [here](https://github.com/gravitational/gravity/tree/dmitri/7.0.x/1588-intermediate-upgrades/lib/update/cluster/versions)

a-palchikov commented on lib/ops/ops.go:
```diff
@@ -1106,16 +1106,7 @@ type SiteOperations []storage.SiteOperation
-		return s.Uninstall.Vars
-	}
-	return storage.OperationVariables{}
+	return (*storage.SiteOperation)(s).Vars()
```
This was an alternative but not identical re-implementation.

a-palchikov commented on tool/gravity/cli/clusterconfig.go:
```diff
@@ -189,10 +189,6 @@ func getConfigUpdater(localEnv, updateEnv *localenv.LocalEnvironment, operation
 	return updater, nil
 }
 
-func (r configInitializer) validatePreconditions(*localenv.LocalEnvironment, ops.Operator, ops.Site) error {
```
Removed as inherent noop - instead initializer can optionally implement a validator interface if required.

bernardjkim commented on Makefile:
```diff
@@ -39,6 +39,8 @@ CURRENT_TAG := $(shell ./version.sh)
 # Abbreviated gravity version to use as a build ID
 GRAVITY_VERSION := $(CURRENT_TAG)
+# Release of the gravity runtime application to builder installer with intermediate steps
+GRAVITY_INTERMEDIATE_RELEASE ?= 6.1.31
```
Will we want to keep this intermediate up to date with the latest 6.1 release? If so we should add an extra comment on when to update this value.

a-palchikov replied:
This is a good question. That entry is mainly for the internal consumption when we build the installer tarball. The version choice is also sort of arbitrary - I went with the latest LTS release at the time I started working on it but there's no compelling reason to always use the latest.
I'll try to formulate a requirement for this as a comment.

bernardjkim commented on lib/builder/builder.go:
```diff
@@ -537,3 +590,65 @@ type GetRepositoryFunc func(*Builder) (string, error)
+	return generateManifest(chart)
+}
+
+func filterUpgradePackageDependencies(packages []pack.PackageEnvelope) (result []pack.PackageEnvelope) {
```
nit: Can we add documentation on what packages are being/should be filtered?

a-palchikov committed:
Fix an issue with dependencies not returning the correct runtime (planet) package
if an alternative image has been configured in global system options.

a-palchikov committed:
Only query the runtime package in finalize

a-palchikov committed:
Address review comments

a-palchikov committed:
Add an upgrade path with an intermediate hop

a-palchikov committed:
Add a new etcd data migration step instead of the previously used backup/restore to workaround an etcd replication issue

a-palchikov committed:
Execute etcd data migration step for each master node in succession. Fix plan builder tests.

a-palchikov committed:
Fix gravity plan resume to properly relay block flag to service

bernardjkim commented on lib/schema/manifest.go:
```diff
@@ -506,9 +513,19 @@ func (d Dependencies) ByName(names ...string) (*loc.Locator, error) {
 	return nil, trace.NotFound("dependencies %q are not defined in the manifest", names)
 }
 
+// Has returns true if dependencies have the specified package
```
Godoc indicates bool return value, but function returns Locator.

a-palchikov replied:
Good catch. Will remove since it is unused.

r0mant commented on lib/builder/syncer.go:
```diff
@@ -70,12 +69,8 @@ func newS3Syncer() (*s3Syncer, error) {
-		Version:    runtimeVersion.String(),
-	})
+func (s *s3Syncer) Sync(ctx context.Context, builder *Builder, runtimeVersion semver.Version) error {
+	tarball, err := s.hub.Get(loc.Gravity.WithVersion(runtimeVersion))
```
Have you made sure that open-source tele build still works with this change? I don't remember if "hub" used here supports "telekube <-> gravity" name translation.

a-palchikov replied:
Good catch - I've changed the application to `telekube` for the oss version so this works.

r0mant commented on lib/localenv/clusterenv.go:
```diff
@@ -71,6 +64,12 @@ func (r *LocalEnvironment) NewClusterEnvironment(opts ...ClusterEnvironmentOptio
+	ctx, cancel := context.WithTimeout(context.TODO(), defaults.AuditLogClientTimeout)
+	defer cancel()
+	auditLog, err := r.AuditLog(ctx)
+	if err != nil && !trace.IsNotFound(err) && !trace.IsAccessDenied(err) {
```
Why is access denied also being filtered out here?

a-palchikov replied:
This does not work on regular nodes:
```
2020-08-04T17:30:00Z WARN [LOCAL]     Failed to create audit log. error:[
ERROR REPORT:
Original Error: *trace.AccessDeniedError access denied to perform action "create" on "user"
Stack Trace:
	/go/_/src/github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/services/role.go:1865 github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/services.RoleSet.CheckAccessToRule
	...
```

r0mant commented on lib/update/cluster/phases/bootstrap.go:
```diff
@@ -18,25 +18,233 @@ package phases
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	_, err = p.bootstrap.Packages.UpsertPackage(resp.Locator, resp.Reader,
```
Is there any particular reason the planet config package is upserted, but the secrets package above is created (with already exists error ignored)?

a-palchikov replied:
Hmm, I don't think there is any specific reason - I will update to use `UpsertPackage` API for the secrets too.

r0mant commented on tool/gravity/cli/update_test.go:
```diff
@@ -0,0 +1,87 @@
+package cli
```
License header

r0mant commented on tool/gravity/cli/update_test.go:
```diff
@@ -0,0 +1,87 @@
+	"gopkg.in/check.v1"
+)
+
+func TestCLI(t *testing.T) { check.TestingT(t) }
```
Are tests in tool/ executed as a part of "make test"?

a-palchikov replied:
Yes, but I just noticed that we don't run enterprise package tests - something to fix in future.

a-palchikov committed:
Address review comments

a-palchikov committed:
Bump planet

a-palchikov committed:
Move etcd data directory migration to gravity as multiple etcd upgrade
hops would have required to modify all intermediate versions of planet
with a command that can be simply implemented externally.
Set minimum support version of 5.5.10 to be able to exercise the double
upgrade hop for etcd.

a-palchikov committed:
Rebase on top of latest changes. Bump e

a-palchikov committed:
Update planet and e to releases
